crates/uv-resolver/src/lock.rs

```diff
@@ -459,6 +468,64 @@ impl Lock {
             .expect("valid index for distribution");
         dist
     }
+
+    /// Convert the [`Lock`] to a [`InMemoryIndex`] that can be used for resolution.
+    ///
+    /// Any packages specified to be upgraded will be ignored.
+    pub fn to_index(
+        &self,
+        install_path: &Path,
+        upgrade: &Upgrade,
+    ) -> Result<InMemoryIndex, LockError> {
+        let distributions =
+            FxOnceMap::with_capacity_and_hasher(self.distributions.len(), Default::default());
+        let mut packages: FxHashMap<_, BTreeMap<Version, PrioritizedDist>> =
+            FxHashMap::with_capacity_and_hasher(self.distributions.len(), Default::default());
+
+        for distribution in &self.distributions {
+            // Skip packages that may be upgraded from their pinned version.
+            if upgrade.contains(distribution.name()) {
+                continue;
+            }
+
+            match distribution.id.source {
+                Source::Registry(..) | Source::Git(..) => {}
+                // Skip local and direct URL dependencies, as their metadata may have been mutated
+                // without a version change.
+                Source::Path(..)
+                | Source::Directory(..)
+                | Source::Editable(..)
+                | Source::Direct(..) => continue,
```

Comment:
Nit: `matches!`?

Reply:
I slightly prefer the exhaustive `match` here because it's clear about what we're ignoring and what we're prefilling.